osu.Server.Queues.ScoreStatisticsProcessor/Processors/UserTotalPerformanceProcessor.cs

```diff
@@ -0,0 +1,136 @@
+        {
+        }
+
+        public void ApplyGlobal(SoloScoreInfo score, MySqlConnection conn)
+        {
+        }
+
+        /// <summary>
+        /// Updates a user's stats with their total PP/accuracy.
+        /// </summary>
+        /// <remarks>
+        /// This does not insert the new stats values into the database.
+        /// </remarks>
+        /// <param name="rulesetId">The ruleset for which to update the total PP.</param>
+        /// <param name="userStats">An existing <see cref="UserStats"/> object to update with.</param>
+        /// <param name="connection">The <see cref="MySqlConnection"/>.</param>
+        /// <param name="transaction">An existing transaction.</param>
+        public async Task UpdateUserStatsAsync(UserStats userStats, int rulesetId, MySqlConnection connection, MySqlTransaction? transaction = null)
+        {
+            beatmapStore ??= await BeatmapStore.CreateAsync(connection, transaction);
+            buildStore ??= await BuildStore.CreateAsync(connection, transaction);
+
+            List<SoloScoreWithPerformance> scores = (await connection.QueryAsync<SoloScoreWithPerformance>(
+                $"SELECT `s`.*, `p`.`pp` FROM {SoloScore.TABLE_NAME} `s` "
+                + $"JOIN {SoloScorePerformance.TABLE_NAME} `p` ON `s`.`id` = `p`.`score_id` "
+                + "WHERE `s`.`user_id` = @UserId "
+                + "AND `s`.`ruleset_id` = @RulesetId", new
+                {
+                    UserId = userStats.user_id,
+                    RulesetId = rulesetId
+                }, transaction: transaction)).ToList();
+
+            Dictionary<int, Beatmap?> beatmaps = new Dictionary<int, Beatmap?>();
+
+            foreach (var score in scores)
+            {
+                if (beatmaps.ContainsKey(score.beatmap_id))
+                    continue;
+
+                beatmaps[score.beatmap_id] = await beatmapStore.GetBeatmapAsync(score.beatmap_id, connection, transaction);
+            }
+
+            // Filter out invalid scores.
+            scores.RemoveAll(s =>
+            {
+                // Score must have a valid pp.
+                if (s.pp == null)
+                    return true;
+
+                // Beatmap must exist.
+                if (!beatmaps.TryGetValue(s.beatmap_id, out var beatmap) || beatmap == null)
+                    return true;
+
+                // Given beatmap needs to be allowed to give performance.
+                if (!beatmapStore.IsBeatmapValidForPerformance(beatmap, s.ruleset_id))
+                    return true;
+
+                // Scores with no build were imported from the legacy high scores tables and are always valid.
+                if (s.ScoreInfo.build_id == null)
+                    return false;
+
+                // Performance needs to be allowed for the build.
+                return buildStore.GetBuild(s.ScoreInfo.build_id.Value)?.allow_performance != true;
+            });
```

Comment:
In the process, this PR also handles non-existent builds, as was mentioned elsewhere but I can't find the comment.